panoply/records.py

```diff
@@ -0,0 +1,55 @@
+from datetime import datetime
+from typing import TypedDict, Dict, List
+
+from .errors.exceptions import WrongTypeOrValueError
+
+
+class PanoplyMetadata(TypedDict):
+    resource_id: str
+    timestamp: str
+
+
+class PanoplyRecord(TypedDict):
+    data: Dict
+    metadata: PanoplyMetadata
+
+
+class PanoplyRecordGroup(TypedDict):
+    data: List[Dict]
+    metadata: PanoplyMetadata
+
+
+def to_record(resource, data) -> PanoplyRecordGroup:
+    """ Converts data and resource to RecordGroup object """
+    validate_resource(resource)
+    data = validate_and_transform_data_type(data)
+    timestamp = get_iso_string()
+    record_group = PanoplyRecordGroup(
+        data=data,
+        metadata={
+            'resource_id': resource,
+            'timestamp': timestamp,
+        }
+    )
+    return record_group
+
+
+def validate_resource(resource):
+    if not isinstance(resource, str) or len(resource.strip()) == 0:
+        raise WrongTypeOrValueError("`resource` must be a non empty string")
+
+
+def validate_and_transform_data_type(data):
+    if isinstance(data, list):
+        if data and not isinstance(data[0], dict):
+            raise WrongTypeOrValueError("Objects inside returned list should be of type dict.")
```

Comment:
Why can't we let type hints deal with this for us? Developer should get errors for not returning the correct type.

Reply:
The datalog components expect a json object, so it means the data must be a dict (inside the list). Where do you see this happening and how is it not failing the entire source?

Reply:
@yonivy In read we return list or None. If we return list of strings the datasource collect will not fail but the writer will (because it expects list of jsons). In such case we will get `pipeline: Error loading datalog json: cannot unmarshal string into Go value of type map[string]interface {}`  
It happened in genericAPI as the users configure what data they want to retrieve by themselves. I have already added checking the type of data there but it may happen in other sources as well.

Reply:
This should fail in the extractor. We cannot allow the datalog contract to be broken. If we can't return json data we should fail quickly with a useful error.